### PR TITLE
Feat/plonk update

### PR DESCRIPTION
**Pull request by ThomasPiellard**
# Description

Updating plonk to the latest version https://eprint.iacr.org/archive/2019/953/1660839633.pdf

[wait for linea to redeploy the contract before merging to avoid incompatibility between prover/verifier]

* The folded quotient is appended to the linearisation polynomial
* the opening of the linearisation polynomial corresponds to the constant term of the algebraic relation
* one less opening to do (since the quotient is now opened throughout the opening of the linearisation polynomial)
* no more modular inverse to compute in the algebraic check to verify a plonk proof

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Test suite for plonk unchanged

# benchmark

* for native chain bls12-377 -> bw6-761 we go from 319k to 312k constraints
* for non native (bw6 -> bn254) the number of constraints is unchanged -> something might be off

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ivokub — COMMENTED** [review]
In general, looks good. On first view seems there is some mismatch from the paper but maybe I have overlooked something. For me there is some discrepancy, but it seems to be because in the paper the constant part is computed separately and I'm not fully understanding it. I trust that you had a better look.

I still haven't looked at Solidity part, will continue later.

File: backend/plonk/bls12-377/prove.go [hunk 1]
```diff
@@ -1279,28 +1219,36 @@ func evaluateXnMinusOneDomainBigCoset(domains [2]*fft.Domain) []fr.Element {
 // The Linearized polynomial is:
 //
 // α²*L₁(ζ)*Z(X)
```
[the remark this answers is not in the document]
Comment (ThomasPiellard) [reply backend/plonk/bls12-377/prove.go:1221]: in fact in the linearised polynomial we leave the constant part computation (constant=everything that can be readily computed by the verifier, such as L1(zeta) ) to the verifier, and the opening of the linearised polynomial is supposed to be the constant part, so the -L1(zeta) for instance is computed by the verifier

File: backend/plonk/bls12-377/prove.go [hunk 2]
```diff
@@ -1279,28 +1219,36 @@ func evaluateXnMinusOneDomainBigCoset(domains [2]*fft.Domain) []fr.Element {
...
-	// first part: individual constraints
+
+	// l(ζ)r(ζ)
 	var rl fr.Element
 	rl.Mul(&rZeta, &lZeta)
 
-	// second part:
-	// Z(μζ)(l(ζ)+β*s1(ζ)+γ)*(r(ζ)+β*s2(ζ)+γ)*β*s3(X)-Z(X)(l(ζ)+β*id1(ζ)+γ)*(r(ζ)+β*id2(ζ)+γ)*(o(ζ)+β*id3(ζ)+γ)
+	// s1 =  α*(l(ζ)+β*s1(β)+γ)*(r(ζ)+β*s2(β)+γ)*β*Z(μζ)
+	// s2 = -α*(l(ζ)+β*ζ+γ)*(r(ζ)+β*u*ζ+γ)*(o(ζ)+β*u²*ζ+γ)
```
[the remark this answers is not in the document]
Comment (ThomasPiellard) [reply backend/plonk/bls12-377/prove.go:1234]: so in this the sign doesn't matter, in the older implem that we had for some reason I inverted the sign of the expression and I kept it like that in this version (we check that +- (s1-s2)=0 mod x**n - 1 so it doesn't matter if it's a + or - )

File: backend/plonk/bls12-377/prove.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivokub) [reply backend/plonk/bls12-377/prove.go:1221]: Ah, I understand now. This also applies to https://github.com/Consensys/gnark/pull/1044#discussion_r1558283331 as well I think.

File: backend/plonk/bls12-377/prove.go [hunk 3]
```diff
@@ -1279,28 +1219,36 @@ func evaluateXnMinusOneDomainBigCoset(domains [2]*fft.Domain) []fr.Element {
...
 	// TODO @gbotrel rename
-	// first part: individual constraints
+
+	// l(ζ)r(ζ)
 	var rl fr.Element
 	rl.Mul(&rZeta, &lZeta)
 
-	// second part:
-	// Z(μζ)(l(ζ)+β*s1(ζ)+γ)*(r(ζ)+β*s2(ζ)+γ)*β*s3(X)-Z(X)(l(ζ)+β*id1(ζ)+γ)*(r(ζ)+β*id2(ζ)+γ)*(o(ζ)+β*id3(ζ)+γ)
+	// s1 =  α*(l(ζ)+β*s1(β)+γ)*(r(ζ)+β*s2(β)+γ)*β*Z(μζ)
```
[the remark this answers is not in the document]
Comment (ivokub) [reply backend/plonk/bls12-377/prove.go:1233]: Resolved, this can be computed by the verifier itself.

[the remark this answers is not in the document]
Comment (ThomasPiellard) [reply backend/plonk/bls12-377/prove.go:1233]: the missing part in this expression is part of the constant term of the linearised polynomial that the verifier has to compute (cf step 8 of the verifier algo to see how he computes the missing term of s1)

File: backend/plonk/bn254/solidity.go [hunk 4]
```diff
@@ -914,14 +916,10 @@ contract PlonkVerifier {
...
 
-        mstore(add(state, STATE_FOLDED_DIGESTS_X), mload(add(state, STATE_FOLDED_H_X)))
-        mstore(add(state, STATE_FOLDED_DIGESTS_Y), mload(add(state, STATE_FOLDED_H_Y)))
-        mstore(add(state, STATE_FOLDED_CLAIMED_VALUES), calldataload(add(aproof, PROOF_QUOTIENT_POLYNOMIAL_AT_ZETA)))
-
-        point_acc_mul(state_folded_digests, add(state, STATE_LINEARISED_POLYNOMIAL_X), acc_gamma, mPtr)
-        fr_acc_mul_calldata(add(state, STATE_FOLDED_CLAIMED_VALUES), add(aproof, PROOF_LINEARISED_POLYNOMIAL_AT_ZETA), acc_gamma)
+        mstore(add(state, STATE_FOLDED_DIGESTS_X), mload(add(state, STATE_LINEARISED_POLYNOMIAL_X)))
+        mstore(add(state, STATE_FOLDED_DIGESTS_Y), mload(add(state, STATE_LINEARISED_POLYNOMIAL_Y)))
+        mstore(add(state, STATE_FOLDED_CLAIMED_VALUES), calldataload(add(aproof, PROOF_LINEARISED_POLYNOMIAL_AT_ZETA)))
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:922]: With these changes, do we need to change the function documentation? Now `H` is part of `Linearized_polynomial`?

Comment (ivokub) [reply backend/plonk/bn254/solidity.go:922]: And also the exponents of `gamma` in the method documentation should be one less for the rest?

File: backend/plonk/bls12-377/verify.go [hunk 5]
```diff
@@ -137,133 +143,127 @@ func Verify(proof *Proof, vk *VerifyingKey, publicWitness fr.Vector, opts ...bac
...
+	// , corresponding to the value of the linearisation polynomiat at ζ
+	// PI(ζ) - α²*L₁(ζ) + α(l(ζ)+β*s1(ζ)+γ)(r(ζ)+β*s2(ζ)+γ)(o(ζ)+γ)*z(ωζ)
+	var constLin fr.Element
+	constLin.Mul(&beta, &s1).Add(&constLin, &gamma).Add(&constLin, &l)       // (l(ζ)+β*s1(ζ)+γ)
+	tmp.Mul(&s2, &beta).Add(&tmp, &gamma).Add(&tmp, &r)                      // (r(ζ)+β*s2(ζ)+γ)
+	constLin.Mul(&constLin, &tmp)                                            // (l(ζ)+β*s1(ζ)+γ)(r(ζ)+β*s2(ζ)+γ)
+	tmp.Add(&o, &gamma)                                                      // (o(ζ)+γ)
+	constLin.Mul(&tmp, &constLin).Mul(&constLin, &alpha).Mul(&constLin, &zu) // α(l(ζ)+β*s1(ζ)+γ)(r(ζ)+β*s2(ζ)+γ)(o(ζ)+γ)*z(ωζ)
+
+	constLin.Sub(&constLin, &alphaSquareLagrangeOne).Add(&constLin, &pi) // PI(ζ) - α²*L₁(ζ) - α(l(ζ)+β*s1(ζ)+γ)(r(ζ)+β*s2(ζ)+γ)(o(ζ)+γ)*z(ωζ)
```
[the remark this answers is not in the document]
Comment (ThomasPiellard) [reply backend/plonk/bls12-377/verify.go:192]: you are right it's not valid the alpha part should not be negated

File: backend/plonk/bn254/solidity.go [hunk 6]
```diff
@@ -981,48 +975,43 @@ contract PlonkVerifier {
...
-        mstore(add(mPtr,0x1e0), VK_S2_COM_Y)
+        mstore(add(mPtr,0x40), mload(add(state, STATE_LINEARISED_POLYNOMIAL_X)))
+        mstore(add(mPtr,0x60), mload(add(state, STATE_LINEARISED_POLYNOMIAL_Y)))
+        calldatacopy(add(mPtr, 0x80), add(aproof, PROOF_L_COM_X), 0xc0)
+        mstore(add(mPtr,0x140), VK_S1_COM_X)
+        mstore(add(mPtr,0x160), VK_S1_COM_Y)
+        mstore(add(mPtr,0x180), VK_S2_COM_X)
+        mstore(add(mPtr,0x1a0), VK_S2_COM_Y)
+        
+        let offset := 0x1c0
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:988]: The in-template computation of `offset` computation at the top of the template seems nice and more future-proof, but I guess changing it here (and in the rest of template) would be too significant change, so can keep as is.

File: backend/plonk/bls12-377/verify.go [hunk 7]
```diff
@@ -137,133 +143,127 @@ func Verify(proof *Proof, vk *VerifyingKey, publicWitness fr.Vector, opts ...bac
...
+	_s1.Mul(&_s1, &tmp).Mul(&_s1, &beta).Mul(&_s1, &alpha).Mul(&_s1, &zu) // α*(l(ζ)+β*s1(β)+γ)*(r(ζ)+β*s2(β)+γ)*β*Z(μζ)
+
+	// _s2 = -α*(l(ζ)+β*ζ+γ)*(r(ζ)+β*u*ζ+γ)*(o(ζ)+β*u²*ζ+γ)
+	_s2.Mul(&beta, &zeta).Add(&_s2, &gamma).Add(&_s2, &l)                                                     // (l(ζ)+β*ζ+γ)
+	tmp.Mul(&beta, &vk.CosetShift).Mul(&tmp, &zeta).Add(&tmp, &gamma).Add(&tmp, &r)                           // (r(ζ)+β*u*ζ+γ)
+	_s2.Mul(&_s2, &tmp)                                                                                       // (l(ζ)+β*ζ+γ)*(r(ζ)+β*u*ζ+γ)
+	tmp.Mul(&beta, &vk.CosetShift).Mul(&tmp, &vk.CosetShift).Mul(&tmp, &zeta).Add(&tmp, &o).Add(&tmp, &gamma) // (o(ζ)+β*u²*ζ+γ)
+	_s2.Mul(&_s2, &tmp).Mul(&_s2, &alpha).Neg(&_s2)                                                           // -α*(l(ζ)+β*ζ+γ)*(r(ζ)+β*u*ζ+γ)*(o(ζ)+β*u²*ζ+γ)
+
+	// α²*L₁(ζ) - α*(l(ζ)+β*ζ+γ)*(r(ζ)+β*u*ζ+γ)*(o(ζ)+β*u²*ζ+γ)
```
[the remark this answers is not in the document]
Comment (ThomasPiellard) [reply backend/plonk/bls12-377/verify.go:222]: mmh where exactly ?

[the remark this answers is not in the document]
Comment (ivokub) [reply backend/plonk/bls12-377/verify.go:222]: I'm thinking about this part.

image

File: backend/plonk/bn254/solidity.go [hunk 8]
```diff
@@ -1170,49 +1169,52 @@ contract PlonkVerifier {
...
+        
         let state := mload(0x40)
 
         // (l(ζ)+β*s1(ζ)+γ)
-        let s1 := add(mload(0x40), STATE_LAST_MEM)
-        mstore(s1, mulmod(calldataload(add(aproof, PROOF_S1_AT_ZETA)), mload(add(state, STATE_BETA)), R_MOD))
-        mstore(s1, addmod(mload(s1), mload(add(state, STATE_GAMMA)), R_MOD))
-        mstore(s1, addmod(mload(s1), calldataload(add(aproof, PROOF_L_AT_ZETA)), R_MOD))
+        let s1
+        s1 := mulmod(calldataload(add(aproof, PROOF_S1_AT_ZETA)), mload(add(state, STATE_BETA)), R_MOD)
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:1187]: Is there a difference between using `s1 := mulmod(...)` vs `mstore(s1, mulmod(...))`?

**Review by ivokub — COMMENTED** [review]
Also checked Solidity verifier. It is a lot better now (with the changes and using more constants for precompile targets and in-template offset computation). Imo the only thing which may be a bit confusing is that we still have separate function for `fold_h` where natively compute inside linearized polynomial computation. But imo it is clearer this way and we have shallower call stack, so I would keep as is.

**Review by ivokub — APPROVED** [review]
Looks good from my side after the fixes also.

File: backend/plonk/bls12-377/verify.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ivokub) [reply backend/plonk/bls12-377/verify.go:222]: Ah, I think I see it now. u is the folding coefficient and computed inside kzg.FoldProof? I had misunderstood it.

File: backend/plonk/bn254/solidity.go [hunk 9]
```diff
@@ -1395,8 +1385,8 @@ func (proof *Proof) MarshalSolidity() []byte {
 	res = append(res, tmp32[:]...)
 
 	// uint256 linearization_polynomial_at_zeta;
-	tmp32 = proof.BatchedProof.ClaimedValues[0].Bytes()
-	res = append(res, tmp32[:]...)
+	// tmp32 = proof.BatchedProof.ClaimedValues[0].Bytes()
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:1388]: Remove comment.

**Review by ivokub — APPROVED** [review]
The Solidity optimization seems good. At first I thought it would be good to also have the same in the in-circuit and native verifiers, but it would make the verifier a bit ugly as would have to manually construct `kzg.BatchProof` with computed constant part of linearized polynomial. The perf diff for native verifier is minimal and for in-circuit verifier I think would be less than 1k constraints for non-native verifier (can omit range-checking the proof elements `proof.BatchProof.ClaimedValues[0]` and can omit a single non-native `AssertIsEqual`).

**Comment by ivokub** [comment]
Hmm, something is also off - BW6 in BN254 is 1131765 constraints *more* with the new version. I guess the additional hashing and folding is more expensive than the saved scalar multiplication.

**Comment by gbotrel** [comment]
> Hmm, something is also off - BW6 in BN254 is 1131765 constraints _more_ with the new version. I guess the additional hashing and folding is more expensive than the saved scalar multiplication.

@yelhousni may investigate; but perf wise, if the finalization circuit is 41M or 42M constraint, doesn't matter we still pay for 64M price.

**Comment by ivokub** [comment]
> Hmm, something is also off - BW6 in BN254 is 1131765 constraints _more_ with the new version. I guess the additional hashing and folding is more expensive than the saved scalar multiplication.

Hmm, I think one place where we add constraints is that previously when we folded H separately, we were able to do 2-MSM and one addition (H0 + zeta^(n+2) H1 + zeta^m(n+2) H2), but now when doing together with rest of linearized polynomial we're doing computing scalar mul for H0. But now we have more hashing as kzg proof folding has more inputs.

**Comment by ivokub** [comment]
> > Hmm, something is also off - BW6 in BN254 is 1131765 constraints _more_ with the new version. I guess the additional hashing and folding is more expensive than the saved scalar multiplication.
> 
> @yelhousni may investigate; but perf wise, if the finalization circuit is 41M or 42M constraint, doesn't matter we still pay for 64M price.

It is 34949956 now vs 33818191 previously.

**Comment by ivokub** [comment]
I retained the in-circuit version which folds H before computing big MSM in linearized polynomial computation. This allows to  avoid one scalar multiplication as before. Now the benchmarks are:
* BLS12 in BW6 master 367119, now **359241**
* BW6 in BN254 master 33818191, now **33871880**

There is still net increase in the number of constraints, but I guess this has to do that for MSM we internally use 2-by-2 MSM and with change of input nb oddity we fall into different case. In case with complete arithmetic (as the test circuits are defined with), we get suboptimal results. But indeed, we can keep rest of optimization for later.

Just for a remark -- with incomplete arithmetic BW6-in-BN254 is now at 29711173, so under 32M threshold.

File: backend/plonk/bn254/solidity.go (hunk 8 above)
Comment (ThomasPiellard) [reply backend/plonk/bn254/solidity.go:1180]: I don't know tbh, I think `s1 := mulmod(...)` uses s1 stored on the (extremely limited) stack while `mstore(s1, mulmod(...))` uses the memory directly so there is no limitation on how many elements we can store (in `mstore(s1, mulmod(...))` s1 is interpreted as an address though it means "store mulmod(...) at address `s1`)

**Comment by yelhousni** [comment]
- BLS12 in BW6 master 367119, now 358515
- BW6 in BN254 master 33818191, now 33885256

**Comment by yelhousni** [comment]
> After [c73574c](https://github.com/Consensys/gnark/pull/1044/commits/c73574cf12e2935c34aa8cd97828a36b2a064516):
> 
> * BLS12 in BW6 master 367119, now ~358515~ **358153** scs.
> * BW6 in BN254 master 33818191, now ~33885256~ ~33884909~ **33883478** scs (after [747d11f](https://github.com/Consensys/gnark/pull/1044/commits/747d11f83c8185a5488de38597521eb43b0b336e)).

* BLS12 in BW6 master 367119, now **358153** scs.
* BW6 in BN254 master 33818191, now **32145464** scs.

**Review by yelhousni — APPROVED** [review]
Good to merge on my side 👍

**Comment by gbotrel** [comment]
can we merge?

**Comment by Sickies** [comment]
Duplicado de #1

**Comment by Sickies** [comment]
[Uploading BTC_USD-202405.csv…]()

**Pull request merged**